File: csharp/ExtractXml.cs
// Extract the raw factur-x.xml payload (UN/CEFACT Cross-Industry Invoice)
// from a Factur-X PDF/A-3 using the InvoiceXML REST API.
//
// Useful when you need to forward the structured XML directly to an existing
// UBL or CII processing pipeline, EDI partner, or archival system.
//
// Get your API key at: https://www.invoicexml.com/account/authentication
// Full API reference:  https://www.invoicexml.com/docs/api/extract/xml
//
// Required NuGet package:
//   dotnet add package Flurl.Http

using Flurl.Http;

public static class ExtractXml
{
    private const string Endpoint = "https://api.invoicexml.com/v1/extract/xml";

    public static async Task<string> RunAsync(
        string pdfPath = "invoice.pdf",
        string outputPath = "factur-x.xml")
    {
        // Get an InvoiceXML API key at: https://www.invoicexml.com/account/authentication
        var apiKey = "INVOICEXML_API_KEY";

        if (!File.Exists(pdfPath))
            throw new FileNotFoundException($"Input PDF not found: {pdfPath}");

        try
        {
            // POST the PDF and receive the raw factur-x.xml content
            // (Cross-Industry Invoice syntax). InvoiceXML handles PDF/A-3
            // attachment extraction and AFRelationship lookup for you.
            var xml = await Endpoint
                .WithOAuthBearerToken(apiKey)
                .PostMultipartAsync(mp => mp
                    .AddFile("file", pdfPath, "application/pdf"))
                .ReceiveString();

            await File.WriteAllTextAsync(outputPath, xml);
            Console.WriteLine($"Factur-X XML extracted to {outputPath} ({xml.Length:N0} chars)");
            return xml;
        }
        catch (FlurlHttpException ex)
        {
            var body = await ex.GetResponseStringAsync();
            Console.Error.WriteLine($"InvoiceXML API error {(int?)ex.StatusCode}: {body}");
            throw;
        }
    }
}

File: nodejs/extract-xml.js
// Extract the raw factur-x.xml from a Factur-X PDF using the InvoiceXML API.
// Get API key: https://www.invoicexml.com/account/authentication
// Docs:        https://www.invoicexml.com/docs/api/extract/xml
//
// Requires Node.js 18+. No dependencies.

const fs = require('fs');

const apiKey = 'YOUR_API_KEY';

(async () => {
    const form = new FormData();
    form.append('file', new Blob([fs.readFileSync('invoice.pdf')], { type: 'application/pdf' }), 'invoice.pdf');

    const response = await fetch('https://api.invoicexml.com/v1/extract/xml', {
        method: 'POST',
        headers: { Authorization: `Bearer ${apiKey}` },
        body: form,
    });

    const xml = await response.text();
    if (!response.ok) {
        console.error(`InvoiceXML API error ${response.status}: ${xml}`);
        process.exit(1);
    }

    fs.writeFileSync('factur-x.xml', xml);
    console.log(`Saved factur-x.xml (${xml.length} chars)`);
})();
